geps/gep-2694.md

```diff
@@ -0,0 +1,108 @@
+
+If there's a regex type that needs to be chosen eventually, RE2 is more appropriate.
+Although the number of gateway APIs supporting RE2 and PCRE are similar when looking over them
+at the moment, there are much more compelling reasons for RE2 to be used.
+
+* Kubernetes only accept expressions in ingress that comply with the RE2 engine syntax.
+* RE2 syntax is a subset of that accepted by PCRE. This means that the syntax of RE2 can be parsed by the PCRE.
+
+### Path Modifiers
+
+Both redirects and rewrites would share the same `PathModifier` types:
+
+```go
+
+// HTTPPathModifierType defines the type of path redirect or rewrite.
+type HTTPPathModifierType string
+
+const (
+  ...
+  // This type of modifier indicates that the full path will be replaced
+  // by the specified value. For example, the path `/foo/bar/baz` with a
+  // RegularExpression match of `/foo/(.*)/baz`, the `bar` will be replaced
+  // with other
+  RegularExpressionHTTPPathModifier HTTPPathModifierType = "RegularExpression"
+)
+
+// HTTPPathModifier defines configuration for path modifiers.
+type HTTPPathModifier struct {
+  // Type defines the type of path modifier.
+  //
+  // +kubebuilder:validation:Enum=Absolute;ReplacePrefixMatch;RegularExpression
+  Type HTTPPathModifierType `json:"type"`
+
+  // Substitution defines the HTTP path value to substitute. An empty value ("")
+  // indicates that the portion of the path to be changed should be removed from
+  // the resulting path. For example, a request to ReplacePrefixMatch "/foo/bar"
+  // with a prefix match of "/foo" would be modified to "/bar".
+  // the path `/foo/bar/baz` with a RegularExpression match of `/foo/(.*)/baz`,
+  // the `bar` will be replaced with others.
+  //
+  // +kubebuilder:validation:MaxLength=1024
+  Substitution string `json:"substitution"`
+}
+```
+
+## Portability
+
+Different gateway APIs support different regexes, we need to find a widely supported regex.
+
+| Gateway APIs         | Regex |
+|----------------------|-------|
+| Istio                | RE2   |
+| Google Cloud         | RE2   |
+| Cilium               | RE2   |
+| Envoy Gateway        | RE2   |
+| Contour              | RE2   |
+| Emissary Ingress     | RE2   |
+| Gloo Edge            | RE2   |
+| Hashicorp Consul     | RE2   |
+| Kuma                 | RE2   |
+| HAProxy              | PCRE  |
+| NGINX                | PCRE  |
+| KONG                 | PCRE  |
+| Traefik              | PCRE  |
+| HAProxy Ingress      | PCRE  |
+| Flomesh Service Mesh | Perl  |
+| Litespeed            | Perl  |
+
+## Alternatives
+
+In the realm of regex, there is no way to perfectly unify regex specific implementations from
+what is available. According to the existing way of handling, a better solution is to let users
+decide their own implementation, and not to make restrictions at the API level. For a more
+user-friendly experience, documentation of the type of Different gateway APIs support regexes
+is available (That's acceptable).
```

Comment:
I'm actually inclined to say that we should _start_ with implementation specific and put the original goal (of standardizing on RE2) here in the alternatives considered, the beauty of that is that if we start with `Implementation Specific` we can _always promote to RE2 later if we think we can get everyone on board_, but the opposite is not true: if we start with RE2 we're committed to it long term.